contents/revisionControl/bestPracticesGit.md

```diff
@@ -0,0 +1,313 @@
+
+Read more:
+- [Rewriting History - Atlassian](https://www.atlassian.com/git/tutorials/rewriting-history)
+- [Don't change public history.](https://sethrobertson.github.io/GitBestPractices/#pubonce)
+- [Force "git push" to overwrite remote files - StackOverflow](https://stackoverflow.com/questions/10510462/force-git-push-to-overwrite-remote-files)
+
+---
+
+## Keep Up To Date
+### Working with Remotes
+_Remotes_ refer to versions of the project you are working on that are hosted elsewhere, usually on the Internet. Remotes are very handy for managing collaboration, e.g. if you have to keep your code in sync with the `upstream` branch of the project, or if you need to pull code from someone else which may not be merged yet.
+
+You can have as many remotes as you want, each possibly being read-only or with read/write privileges.
+
+Remotes are managed using the `git remote` command.
+```bash
+# view all remotes
+git remote -v
+
+# add a remote called "upstream" that points to https://github.com/TEAMMATES/teammates
+git remote add upstream https://github.com/TEAMMATES/teammates
+
+# branch off from the master branch of the upstream repository
+git fetch upstream # get data from upstream repo
+git checkout -b your-fancy-branch upstream/master # makes a new branch off the head of upstream/master
+
+# change the URL for the upstream remote from HTTPS to SSH
+git remote set-url upstream git@github.com:TEAMMATES/teammates.git
+
+# remove a remote named "upstream"
+git remote remove upstream
+```
+
+Read more:
+- [Git Basics - Working with Remotes](https://git-scm.com/book/en/v2/Git-Basics-Working-with-Remotes)
+
+### Rebase versus Merging
+| You should... | When ... |
+|---------------|----------|
+| merge | you created a branch to develop a feature, and now you want these changes to be inside `master` |
+| rebase | you created a branch from `master` to develop a feature, and someone else pushed a change to `master` before you finished |
+
+It is generally considered good practice to rebase your feature branch onto whatever branch you're trying to patch _before_ you make the pull request, resolving any conflicts that arise. This:
+- [keeps the history clear and linear](http://www.bitsnbites.eu/a-tidy-linear-git-history/)
+    - makes backtracking easier
+    - easy to follow history
+    - reverting/rolling back is much simpler
+    - you can use `git bisect` to find regressions on your branch easily without involving unrelated changes from `master`
+- ensures your changes are compatible with the head of the branch you're patching
+- makes reviewing/testing easier by [not including irrelevant code by merging](https://lwn.net/Articles/328436/)
+
+Read more:
+- [A Tidy, Linear Git History](http://www.bitsnbites.eu/a-tidy-linear-git-history/) _this is an excellent article which has formed my opinion on rebase vs. merge_
```

Comment:
It is good to give an overview of each resource like you have done here. Try to do it more. But may be not use italics?

Reply:
I wasn't sure what was the best way to do this, so I took out the italics and uses a hyphen to delimit the start of the overview. I only give an overview for the more important/interesting resources, though. Some like the StackOverflow threads are self-explanatory.